osm2streets/src/output.rs

```diff
@@ -0,0 +1,264 @@
+use itertools::Itertools;
+
+use crate::{Direction, LaneType, StreetNetwork};
+use geo::MapCoordsInPlace;
+use geom::{Distance, Line, PolyLine, Pt2D};
+
+use LaneType::*;
+use SurfaceMaterial::*;
+
+pub struct Surface {
+    pub area: geo::Polygon,
+    pub material: SurfaceMaterial,
+}
+
+pub struct PaintArea {
+    pub area: geo::Polygon,
+    pub color: PaintColor,
+}
+
+impl PaintArea {
+    pub fn new(area: geo::Polygon) -> Self {
+        Self {
+            area,
+            color: PaintColor::White,
+        }
+    }
+    pub fn with_color(area: geo::Polygon, color: PaintColor) -> Self {
+        Self { area, color }
+    }
+}
+
+impl StreetNetwork {
+    /// Generates polygons covering the road, cycle and footpath areas.
+    pub fn get_surfaces(&self) -> Vec<Surface> {
+        let mut output = Vec::new();
+
+        // Add polygons for road surfaces.
+        for road in self.roads.values() {
+            // Generate an area for each contiguous group of footpath, bike path and road lanes.
+            let center_offset = road.total_width() / 2.0;
+            let mut processed_width = Distance::ZERO;
+            for (material, lanes) in road
+                .lane_specs_ltr
+                .iter()
+                .group_by(|l| material_from_lane_type(l.lt))
+                .into_iter()
+            {
+                if let Some(material) = material {
+                    let mut width = Distance::ZERO;
+                    for lane in lanes {
+                        width += lane.width;
+                    }
+                    output.push(Surface {
+                        area: road
+                            .center_line
+                            .shift_either_direction(processed_width - center_offset + width / 2.0)
+                            .unwrap_or_else(|_| road.center_line.clone())
+                            .make_polygons(width)
+                            .into(),
+                        material,
+                    });
+                    processed_width += width;
+                } else {
+                    for lane in lanes {
+                        processed_width += lane.width;
+                    }
+                }
+            }
+        }
+
+        // Polygon per intersection
+        for intersection in self.intersections.values() {
+            // TODO dissect the area into pieces based on movements, like with Roads.
```

Comment:
Very curious what you have in mind here. Literally splitting the polygon by all the movement polylines?

Reply:
Also how we know what material it's made of, so intersections between e.g. paths naturally have material inherited from the roads going through it, instead of asphalt